openhands/runtime/remote/runtime.py

```diff
@@ -83,59 +81,93 @@ def __init__(
         self.runtime_url: str | None = None
 
-        self.instance_id = (
-            sid + str(uuid.uuid4()) if sid is not None else str(uuid.uuid4())
+        self.instance_id = sid
+
+        self._start_or_attach_to_runtime(plugins)
+
+        # Initialize the eventstream and env vars
+        super().__init__(
+            config, event_stream, sid, plugins, env_vars, status_message_callback
         )
-        self.container_name = 'oh-remote-runtime-' + self.instance_id
-        if self.config.sandbox.runtime_container_image is not None:
-            logger.info(
-                f'Running remote runtime with image: {self.config.sandbox.runtime_container_image}'
-            )
-            self.container_image = self.config.sandbox.runtime_container_image
+
+    def _start_or_attach_to_runtime(self, plugins: list[PluginRequirement] | None):
+        existing_runtime = self._check_existing_runtime()
+        if existing_runtime:
+            logger.info(f'Using existing runtime with ID: {self.runtime_id}')
         else:
-            logger.info(
-                f'Building remote runtime with base image: {self.config.sandbox.base_container_image}'
-            )
-            logger.debug(f'RemoteRuntime `{sid}` config:\n{self.config}')
+            self.send_status_message('STATUS$STARTING_CONTAINER')
+            if self.config.sandbox.runtime_container_image is None:
+                logger.info(
+                    f'Building remote runtime with base image: {self.config.sandbox.base_container_image}'
+                )
+                self._build_runtime()
+            else:
+                logger.info(
+                    f'Running remote runtime with image: {self.config.sandbox.runtime_container_image}'
+                )
+            self._start_runtime(plugins)
+        assert (
+            self.runtime_id is not None
+        ), 'Runtime ID is not set. This should never happen.'
+        assert (
+            self.runtime_url is not None
+        ), 'Runtime URL is not set. This should never happen.'
+        self.send_status_message('STATUS$WAITING_FOR_CLIENT')
+        self._wait_until_alive()
+
+    def _check_existing_runtime(self) -> bool:
+        try:
             response = send_request(
                 self.session,
                 'GET',
-                f'{self.config.sandbox.remote_runtime_api_url}/registry_prefix',
-            )
-            response_json = response.json()
-            registry_prefix = response_json['registry_prefix']
-            os.environ['OH_RUNTIME_RUNTIME_IMAGE_REPO'] = (
-                registry_prefix.rstrip('/') + '/runtime'
-            )
-            logger.info(
-                f'Runtime image repo: {os.environ["OH_RUNTIME_RUNTIME_IMAGE_REPO"]}'
+                f'{self.config.sandbox.remote_runtime_api_url}/runtime/{self.instance_id}',
             )
+            if response.status_code == 200:
+                self._parse_runtime_response(response)
```

Comment:
Should we also check for pod `status`? e.g., when pod status is failure/error, we should restart instead of re-use it.

Reply:
Or we could do it on the server end -- return 404 when pod is failed/error.

Reply:
Yes--I will follow up with this now that I have pause/resume on the way